[clang] Support __typeof_unqual__ in all C modes

nathanchance opened the pull request:
GCC has added `__typeof_unqual__` to allow `typeof_unqual` to be used in all C modes (not just C23 and newer), similar to `__typeof__` and `typeof`.

https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=607d9d50ee44163cee621cd991600acaf78c2fee

The Linux kernel would like to start using `__typeof_unqual__` to strip type qualifiers such as address spaces from inputs to macros but cannot switch to C23 due to compiler version requirements.

Match GCC and allow `__typeof_unqual__` in all C modes.

Closes: https://github.com/llvm/llvm-project/issues/76423
Link: https://lore.kernel.org/CAFULd4YG21NdF_qNVBGDtXO6xnaYFeRPvKicB=gpgUUqYE=4jw@mail.gmail.com/

PR notes/commentary:

- I am not sure if there are any other tests I should add, I could not tell where `__typeof__` was being used normally versus being explicitly tested aside from this test.
- If there are any comments in the source code that should be updated, I am more than happy to do so, I just didn't want to be preemptive if they were not going to matter.

nathanchance committed:
[clang] Support __typeof_unqual__ in all C modes

GCC has added __typeof_unqual__ to allow typeof_unqual to be used in all
C modes (not just C23 and newer), similar to __typeof__ and typeof.

https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=607d9d50ee44163cee621cd991600acaf78c2fee

The Linux kernel would like to start using __typeof_unqual__ to strip
type qualifiers such as address spaces from inputs to macros but cannot
switch to C23 due to compiler version requirements.

Match GCC and allow __typeof_unqual__ in all C modes.

Closes: https://github.com/llvm/llvm-project/issues/76423
Link: https://lore.kernel.org/CAFULd4YG21NdF_qNVBGDtXO6xnaYFeRPvKicB=gpgUUqYE=4jw@mail.gmail.com/
Signed-off-by: Nathan Chancellor <nathan@kernel.org>

nickdesaulniers commented on clang/test/Parser/c2x-typeof-ext-warns.c:
```diff
@@ -12,9 +12,10 @@
+// Show what happens with the underscored version of the keywords, which are
+// conforming extensions.
 __typeof__(int) i = 12;
+__typeof_unqual__(int) u = 12;
```
consider adding a test for the non-underscore-suffixed variant as well.

nathanchance replied:
Should I add one for `__typeof` as well?

AaronBallman replied:
No harm in adding that extra coverage as well.

nathanchance replied:
Done in 6186b113e9c75c808ff0a08455353d2bb8c3f50b.

nickdesaulniers commented on clang/include/clang/Basic/TokenKinds.def:
```diff
@@ -665,28 +665,30 @@ KEYWORD(__kindof                     , KEYOBJC)
+ALIAS("__signed__"       , signed       , KEYALL)
+ALIAS("__typeof"         , typeof       , KEYALL)
+ALIAS("__typeof__"       , typeof       , KEYALL)
+ALIAS("__typeof_unqual"  , typeof_unqual, KEYNOCXX)
+ALIAS("__typeof_unqual__", typeof_unqual, KEYNOCXX)
```
Why not c++?

nathanchance replied:
No real reason other than matching GCC's behavior as documented in that commit message. Is `typeof_unqual` allowed in C++? It appears to be C only in GCC but I could not seem to find anything conclusive in Clang.

AaronBallman replied:
I think it's reasonable to expose this in C++ as well since it's a conforming extension (it should Just Work™).

nathanchance replied:
Done in 960cf9bd3a7252d93119f02c2600891d9d7183ad.

frederick-vs-ja replied:
> No real reason other than matching GCC's behavior as documented in that commit message. Is `typeof_unqual` allowed in C++? It appears to be C only in GCC but I could not seem to find anything conclusive in Clang.

The paper for adding `typeof` and `typeof_unqual` to C++ seems rejected currently (cplusplus/papers#1630).
But I do believe `typeof_unqual` is useful in C++ as it can improve compiler throughput in cases like `std::remove_cvref_t<decltype((e))>`.

h-vetinari replied:
The C++ rejection should be questioned IMO. To my understanding, once upon a time C++ went with `decltype` to not steal `typeof` from C, waiting for WG14 to standardize it (and once that happened - admittedly decades later -, it was [explicitly](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2930.pdf) made more compatible with C++ during that process). Now that it's finally there, clang could reasonably expose this in all C/C++ modes, also without the leading underscores.

nickdesaulniers reviewed (commented):
Thanks for the patch!

AaronBallman reviewed (commented):
Thank you for this; I think it's a very reasonable extension!

nathanchance committed:
fixup! [clang] Support __typeof_unqual__ in all C modes

Allow __typeof_unqual__ in C++

Signed-off-by: Nathan Chancellor <nathan@kernel.org>

nathanchance committed:
fixup! [clang] Support __typeof_unqual__ in all C modes

While in the area, add a test for __typeof, which should be handled in
the same manner as __typeof__.

Signed-off-by: Nathan Chancellor <nathan@kernel.org>

AaronBallman reviewed (approved):
LGTM, please add a C++-specific test when landing, since this is also going into C++.

nathanchance committed:
fixup! [clang] Support __typeof_unqual__ in all C modes

Add a C++ test, as this will be exposed there as well

Signed-off-by: Nathan Chancellor <nathan@kernel.org>

nathanchance commented:
> LGTM, please add a C++-specific test when landing, since this is also going into C++.

Does 4f9e79bafdd2fb9d1414e474367241244bffe29d seem reasonable on that front? Should I add a test that `typeof_unqual` is not accepted for C++? As far as I understand it, that is not supported by any compiler or the standard, correct?

AaronBallman commented:
> > LGTM, please add a C++-specific test when landing, since this is also going into C++.
> 
> Does [4f9e79b](https://github.com/llvm/llvm-project/commit/4f9e79bafdd2fb9d1414e474367241244bffe29d) seem reasonable on that front? 

Yup, thank you!

> Should I add a test that `typeof_unqual` is not accepted for C++? As far as I understand it, that is not supported by any compiler or the standard, correct?

I don't think there's a need for that (your patch doesn't change the status quo for `typeof_unqual`). That said, I will rarely complain about additional test coverage. ;-)

nathanchance committed:
fixup! [clang] Support __typeof_unqual__ in all C modes

Include more coverage for new test

Signed-off-by: Nathan Chancellor <nathan@kernel.org>

nathanchance commented:
> > Should I add a test that `typeof_unqual` is not accepted for C++? As far as I understand it, that is not supported by any compiler or the standard, correct?
> 
> I don't think there's a need for that (your patch doesn't change the status quo for `typeof_unqual`). That said, I will rarely complain about additional test coverage. ;-)

Since it is a new test, I will add that coverage just to make it a little more obvious that is expected: c89b9694c8dc05b1416d2076a067316ed19f0049

I'll merge this once presubmit is finished and no obvious regressions are noticed, thank you both for the quick feedback!

nathanchance commented:
Would it make sense to apply this to `release/18.x` to get it into the hands of users quicker? It seems like it should be a safe change to apply. It sounds like GCC 14 will be the first one to support `__typeof_unqual__`, which should be released quite soon if I understand their release cycle correctly, so it makes sense that Clang would support it around the same time. No worries if it does not seem reasonable, just figured it was worth asking.

AaronBallman commented:
I agree that the feature is safe, but I don't think it qualifies because it's adding a new feature rather than fixing an issue. That said, if @tru or @tstellar think it's worth adding as release managers, I won't block it (because it really should be quite safe).

tru commented:
The final decision is Toms, but I don't think it qualifies since we are so late in the current process and that 19 will start in just a few months.
